Introduce TestSubscriber public API to reactor-test

simonbasle opened the pull request:
Fixes #2692.

~Note: this PR depends on #2707 and will need to be retargeted to `main` once it is merged.~

rstoyanchev commented on reactor-test/src/main/java/reactor/test/subscriber/TestSubscriber.java:
```diff
@@ -0,0 +1,417 @@
+		return new ArrayList<>(this.receivedOnNext);
+	}
+
+	public List<T> getOnNextSublistReceivedAfterCancel() {
```
`getReceivedOnNextAfterCancellation` aligning with `getReceivedOnNext`?

simonbasle replied:
this can probably be conveyed through the javadoc, but one thing I want to point out here is that this is (conceptually) a sublist of the `getReceivedOnNext()` list. ie. the later always contains the former. this method is offered for convenience in case one wants to check how the source behaves wrt cancellation, namely: can it produce "extraneous" elements after cancellation (technically allowed by the RS spec).

rstoyanchev reviewed (commented):
Looks good overall with some comments below. Also, it would be good to have some Javadoc on the public accessors. For `getProtocolErrors`, for example, I wasn't entirely sure until searched through the usages.

simonbasle commented on reactor-test/src/main/java/reactor/test/subscriber/TestSubscriber.java:
```diff
@@ -0,0 +1,417 @@
+		 */
+		NO_REQUIREMENT;
+
+	}
```
got suggestion

simonbasle commented on reactor-test/src/main/java/reactor/test/subscriber/TestSubscriber.java:
```diff
@@ -0,0 +1,417 @@
+
+/**
+ * A {@link CoreSubscriber} that can be attached to any {@link org.reactivestreams.Publisher} to later assert which
+ * events occurred at runtime. This is an alternative to {@link reactor.test.StepVerifier} which allows to evaluate
+ * more complex scenarios where there are more than one possible outcome (eg. racing).
```
I'll rework the head javadoc that has other elements that are outdated and include your suggestion 👍

simonbasle commented:
I was wondering about failure conditions and how to "access" them. There are conceptually 3 sources of exceptions:
 1. the natural `onError` from source `Publisher`
 2. "protocol errors", ie the `Publisher` behaves badly and sends a sequence of signals that doesn't respect the Reactive Streams spec
 3. illegal state that is not directly a RS violation, e.g. fusion is requested but not correctly established. Arguably, reusing the `TestSubscriber` could be considered both 2) and 3)

So for `onError` we're in a good place:
 - stored as `terminalSignal`
 - retrievable by `getTerminalSignal()`, `expectTerminalSignal()`, `expectTerminalError()`
 - doesn't cause `block()` to throw

For protocol errors I think we're also good:
 - stored and accessible in `getProtocolErrors()`
 - doesn't cause `block()` to throw
 - most of the error we detect, to the exception of double subscription, imply the publisher has already terminated

For illegal state, not so sure:
 - stored in `internalFailure`
 - causes a cancellation of subscription when it happens
 - no particular accessor
 - DO causes `block()` to throw

Thus so far the only way to surface these is to call `block()`, which will immediately throw these exceptions (as `AssertionError` in fact).
I think we'd need an accessor too, but I wonder if that would make the number of places to look for exceptions too high (3, basically). On the other hand, if we were to add these to `getProtocolErrors()`, it wouldn't be straightforward to distinguish between failures defined in the RS spec vs failures linked to the TestSubscriber's configuration...


 - [ ] Does it make sense to throw in block() only for category 3?
 - [ ] If so, should we have a separate accessor for category 3?
 - [ ] Should we consider mixing 2) and 3) into a single category (in which case protocol errors would interrupt `block()` as well)?
 - [ ] Should we have an hybrid approach were only 3) interrupt the block(), but they also get added as protocol errors?

cc @reactor/team, @rstoyanchev

simonbasle commented:
I've added some tests and added all the necessary javadoc @rstoyanchev.
I still need to cover `block()` and `scan` methods in tests, but this is almost shaped up.

Regarding the above comment, I went the "store separately and access separately" way (third accessor) + added an `isDone()` overarching getter and a `isFailed()` specific getter for that case.

rstoyanchev commented:
@simonbasle, regarding exceptions, it sounds okay to me that "internalFailure" is simply raised on `block()` without the need for an accessor. It i illegal state after all and by definition it shouldn't be common. I'm wondering though if other accessors shouldn't also raise an error too like `block()`, e.g. if you call `isTerminated` or `isCancelled`, the `expect*` methods, etc. So however you use `TestSubscriber`, you don't have to check for illegal state, but it also doesn't remain invisible.

For terminal signals clearly those would be of primary interest in most often. I'm less clear on protocol errors. Does it really make sense to check in every test, or wouldn't it be more useful to be able to run in some "strict" mode in which protocol errors would also become visible in some way, either causing failure like "internalFailure" or somehow being reported. I haven't thought through how this might work but just thinking about how the feature might be used.

simonbasle commented:
At first I wasn't sure about raising the `internalFailure` in getters, but I think it is definitely a good idea to rise it in both `expectXxx` methods.

`isXxx` getters are expected to be callable at any point in a test, ie. returning `false` is not necessarily indicative of a test failure (one could be asserting that at the beginning of the test, the `TestSubscriber` hasn't been yet `isTerminated()` for instance).

That said, illegal states so far all occur at subscription time. Maybe we should rename these "subscription failures"? If the subscription fails, there is no real point in checking for terminal state, received onNext or even protocol errors... So it could make sense to fail the test in all these cases 👍 

For protocol errors, it is a bit different. They represent extraneous unexpected signals that mostly come _after_ a terminal signal. If the test itself doesn't anticipate a protocol error, there's no reliable way for the `TestSubscriber` itself to know that an extraneous signal will come in after that terminal signal it just received. How long should the `TestSubscriber` wait to see if a protocol error happens?

So it is up to the test to be explicitly setting things up with protocol errors. Mostly, I expect the usage to be either:
 - for confirming that a given source is misbehaving 
OR 
 - to work with a misbehaving source + an intermediate operator, and asserting said operator didn't produce any protocol errors (ie. checking an operator protects against protocol errors when the source of said operator is malformed)

simonbasle commented:
since the `AssertSubscriber` is copied in RSocket-java and used mainly in concurrency / racing tests, how much more thread-safety do we want to add to the current state of `TestSubscriber`?

simonbasle commented:
this needs a way to assert signal order as a whole (onNext + onComplete + onError). RSocket tests need that in order to detect eg. `onComplete -> onNext` or `onError -> onComplete`, etc...

OlegDokuka commented on reactor-test/src/main/java/reactor/test/subscriber/TestSubscriber.java:
```diff
@@ -0,0 +1,752 @@
+				internalCancel();
+				throw new IllegalStateException("Request is short circuited in SYNC fusion mode, and should not be explicitly used");
+			}
+			this.s.request(n);
```
should not we call drain loop here if FUSION is ASYNC

OlegDokuka replied:
That is true as well! So lets keep it as is

OlegDokuka reviewed (changes requested):
So far looks good. Need to use it in rsocket first to see how well it fits but noticed few issues that can break expectations

simonbasle commented on reactor-test/src/main/java/reactor/test/subscriber/TestSubscriber.java:
```diff
@@ -0,0 +1,752 @@
+		if (t == null) {
+			if (this.fusionMode == Fuseable.ASYNC) {
+				assert this.qs != null;
+				for (; ; ) {
```
indeed, good catch ! I think `AssertSubscriber` could "overflow" due to that actually 🤕

simonbasle replied:
updated the code 👍

OlegDokuka commented:
Spot a few more things. 
In RSocket-Java we need to:
1. Support requests before subscription established (https://github.com/rsocket/rsocket-java/blob/master/rsocket-core/src/test/java/io/rsocket/internal/UnboundedProcessorTest.java#L312)
2. Cancellation should clear the upstream in case of async fusion (https://github.com/rsocket/rsocket-java/blob/master/rsocket-core/src/test/java/io/rsocket/internal/subscriber/AssertSubscriber.java#L868)

simonbasle commented:
putting this back on draft, will probably need more internal rework to support RSocket-java (drain loop included)

simonbasle commented:
(just rebased on `main`)

simonbasle commented:
@OlegDokuka @rstoyanchev I've made one final modification: I've extracted the `TestSubscriber` and `ConditionalTestSubscriber` interfaces and split the concrete implementations as package-private classes.
